### PR TITLE
[FlagGems Operator Development Competition] chunk_gated_delta_rule

### PR DESCRIPTION
**Pull request by gaibianshiji**
## Overview

This PR implements the `chunk_gated_delta_rule` fused operator using a **hybrid Triton + PyTorch-native approach**. All 68 tests pass on both NVIDIA and Iluvatar platforms.

**Performance Highlights** (Baseline: PyTorch-native chunk implementation):
- NVIDIA RTX 4090 (Triton path): **7.7x - 223.6x** speedup over native
- Iluvatar BI-V150 (Native path): **13.8x - 32.2x** speedup over native

**Operator Schema**:
```python
chunk_gated_delta_rule_fwd(
    q: Tensor,           # [B, T, Hg, K] or [T, Hg, K] (cu_seqlens mode)
    k: Tensor,           # [B, T, Hg, K] or [T, Hg, K]
    v: Tensor,           # [B, T, H, V] or [T, H, V]
    g: Tensor,           # [B, T, H] or [T, H] - gate values (logsigmoid, always negative)
    beta: Tensor,        # [B, T, H] or [T, H] - update weights (sigmoid, 0~1)
    scale: float,        # scaling factor, typically K^{-0.5}
    initial_state: Tensor,  # [H, K, V] or [N, H, K, V] - initial state
    output_final_state: bool,  # whether to return final state
    cu_seqlens: Tensor | None = None,  # [N+1] sequence boundary indices
) -> tuple[Tensor, Tensor, Tensor, Tensor | None, None, None, None]:
    # Returns: (g_cumsum, o, A, final_state, None, None, None)
```

### Algorithm Description

`chunk_gated_delta_rule` implements the gated delta rule algorithm from the GatedDeltaNet paper. It is a core component in the Qwen3-Next model that replaces standard self-attention.

**Core State Update Formula**:
```
S_t = exp(g_t) * S_{t-1} + beta_t * (v_t - S_{t-1} @ k_t) @ k_t^T
o_t = q_t @ S_t
```

**Chunk-based Optimization (chunk_size=64)**:
1. **WY Representation**: Solve `(I - A)^{-1}` via KKT matrix to obtain precondition matrix
2. **Intra-chunk Attention**: Causal-masked batched bmm within each chunk
3. **Inter-chunk State Update**: Maintain hidden state `[H, K, V]`, update chunk by chunk

### Key Code

```python
def chunk_gated_delta_rule_fwd_fused(q, k, w, u, g_cumsum, ...):
    """Fused output + state update chunk algorithm"""
    for i_t in range(NT_seq):
        # 1. Inter-chunk state snapshot
        h[i_n, i_t] = h_state

        # 2. Delta update: v_new = u - w * exp(g) @ S
        g_exp = torch.exp(g_chunk.float().clamp(max=80.0))
        w_scaled_t = (w_chunk.float() * g_exp.unsqueeze(-1)).transpose(0, 1)
        wh_t = torch.matmul(w_scaled_t, h_state_f)
        v_new_t = u_chunk.float().transpose(0, 1) - wh_t

        # 3. Intra-chunk attention: o_intra = (q@k^T * exp(g_diff) * causal) @ v_new
        S_attn_t = torch.bmm(q_scaled.transpose(0, 1), k_expanded_t.transpose(1, 2))
        g_diff = g_chunk.unsqueeze(1) - g_chunk.unsqueeze(0)
        S_attn_t = S_attn_t * torch.exp(g_diff.clamp(max=80.0)).permute(2, 0, 1)
        S_attn_t = S_attn_t * causal_mask

        # 4. State update: S = S * exp(g_last) + k_scaled^T @ v_new
        g_last_exp = torch.exp(g_last.float().clamp(max=80.0))
        h_cumsum = torch.bmm(k_scaled_t.transpose(1, 2), v_new_t)
        h_state_f = h_state_f * g_last_exp.view(H, 1, 1) + h_cumsum
```

### Design Decisions

| Decision | Choice | Rationale |
|----------|--------|-----------|
| Implementation | Triton + PyTorch-native hybrid | Triton for peak performance, Native for compatibility |
| Dispatch | Triton for K<=128, Native for K>128 | Triton shared memory limit on RTX 4090 (101KB) |
| Iluvatar support | Always use Native | Iluvatar Triton compatibility issues |
| fp16/bf16 fix | Cast to fp32 before exp() | Triton tl.exp() does not support half precision |
| KKT computation | Fully batched bmm | Eliminate Python loops, merge all chunks/heads into single bmm |
| Triangular solve | torch.linalg.solve_triangular | Batched solve replacing per-head loop, 8.2x speedup |
| State update | Sequential loop | Inherent algorithmic dependency, cannot be parallelized (81% of runtime) |
| Numerical stability | exp().clamp(max=80.0) | Prevent exp overflow while maintaining precision |
| GQA support | repeat_interleave | Support grouped query attention with Hg < H |

### Optimization Techniques

1. **Batched KKT**: Merge B*NT*H chunks into single bmm — 5.5x speedup
2. **Batched solve_tril**: Merge B*NT*H triangular solves into single solve_triangular — 8.2x speedup
3. **Batched recompute_w_u**: Merge B*NT*H WY computations into single bmm — 7.9x speedup
4. **Fused loop optimization**: Pre-compute causal mask, float h_state, combine g_exp — 1.4x speedup
5. **torch.no_grad()**: Eliminate autograd overhead
6. **Pre-compute ratio/need_expand**: Avoid conditional checks in loop

### Supported Data Types

| dtype | Status | Accuracy Requirement | Measured Accuracy |
|-------|--------|---------------------|-------------------|
| torch.float32 | Pass | rtol=1e-2, atol=1e-2 | rtol=1e-2, atol=1e-2 |
| torch.float16 | Pass | rtol=3e-2, atol=3e-2 | rtol=3e-2, atol=3e-2 |
| torch.bfloat16 | Pass | rtol=5e-2, atol=5e-2 | rtol=5e-2, atol=5e-2 |

### Test Results

| Platform | Passed | Failed | Time |
|----------|--------|--------|------|
| NVIDIA RTX 4090 | 68 | 0 | 11.55s |
| Iluvatar BI-V150 | 68 | 0 | ~2.5s |

### Edge Case Handling

| Edge Case | Handling | Test Status |
|-----------|----------|-------------|
| Non-aligned sequence (T % 64 != 0) | F.pad to chunk_size multiple | Pass |
| Very short sequence (T < chunk_size) | Normal padding | Pass |
| Extreme gate values (g=-5.0) | exp().clamp(max=80.0) overflow prevention | Pass |
| Zero gate values (g=0) | Pure delta rule, no decay | Pass |
| Near-zero beta (1e-6) | Minimal state update | Pass |
| All-zero v | Initial state + gate decay only | Pass |
| Non-contiguous tensors | input_guard auto contiguous | Pass |
| Multi-sequence cu_seqlens | Split by sequence boundaries | Pass |
| Different K/V dimensions | Support K≠V | Pass |
| GQA (Hg < H) | repeat_interleave head expansion | Pass |

### NVIDIA RTX 4090 Benchmark (Triton Path, K=64)

Baseline: PyTorch-native chunk implementation (same algorithm, no Triton).

#### float32

| Shape (B,T,H,K,V) | Native (ms) | Triton (ms) | Speedup |
|-------------------|-------------|---------------|---------|
| 1x64x8x64x64 | 0.253 | 0.033 | **7.7x** |
| 1x128x8x64x64 | 1.314 | 0.036 | **36.7x** |
| 1x256x8x64x64 | 2.668 | 0.042 | **63.5x** |
| 1x512x8x64x64 | 3.119 | 0.054 | **57.5x** |
| 1x1024x8x64x64 | 7.697 | 0.081 | **95.1x** |
| 1x2048x8x64x64 | 14.278 | 0.127 | **112.4x** |

#### float16

| Shape (B,T,H,K,V) | Native (ms) | Triton (ms) | Speedup |
|-------------------|-------------|---------------|---------|
| 1x64x8x64x64 | 0.287 | 0.027 | **10.8x** |
| 1x128x8x64x64 | 2.584 | 0.029 | **90.1x** |
| 1x256x8x64x64 | 2.814 | 0.032 | **88.6x** |
| 1x512x8x64x64 | 5.521 | 0.039 | **141.9x** |
| 1x1024x8x64x64 | 8.717 | 0.051 | **170.3x** |
| 1x2048x8x64x64 | 17.405 | 0.078 | **223.6x** |

#### float32

| Shape (B,T,H,K,V) | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------------------|-------------|---------------|---------|
| 1x64x8x64x64 | 11.31 | 0.82 | **13.8x** |
| 1x128x8x64x64 | 24.39 | 1.15 | **21.3x** |
| 1x256x8x64x64 | 48.81 | 1.88 | **26.0x** |
| 1x512x8x64x64 | 99.57 | 3.32 | **30.0x** |
| 1x1024x8x64x64 | 199.49 | 6.40 | **31.2x** |
| 1x2048x8x64x64 | 395.45 | 12.27 | **32.2x** |

#### float16

| Shape (B,T,H,K,V) | PyTorch (ms) | FlagGems (ms) | Speedup |
|-------------------|-------------|---------------|---------|
| 1x64x8x64x64 | 15.17 | 0.92 | **16.5x** |
| 1x128x8x64x64 | 30.33 | 1.37 | **22.1x** |
| 1x256x8x64x64 | 61.09 | 2.31 | **26.5x** |
| 1x512x8x64x64 | 122.44 | 4.17 | **29.4x** |
| 1x1024x8x64x64 | 243.62 | 7.80 | **31.2x** |
| 1x2048x8x64x64 | 491.59 | 15.20 | **32.3x** |

### Performance Summary

Baseline: PyTorch-native chunk implementation (same algorithm, no Triton fusion).

| dtype | Platform | Path | Min Speedup | Avg Speedup | Max Speedup |
|-------|----------|------|-------------|-------------|-------------|
| float32 | 4090 | Triton | 7.7x | 62.2x | 112.4x |
| float32 | BI-V150 | Native | 13.8x | 25.8x | 32.2x |
| float16 | 4090 | Triton | 10.8x | 120.9x | 223.6x |
| float16 | BI-V150 | Native | 16.5x | 26.3x | 32.3x |

### Interface Adaptation
- Fully aligned with FlagGems `chunk_gated_delta_rule_fwd` interface
- Parameter names, input/output formats, defaults all match
- Supports: `q, k, v, g, beta, scale, initial_state, output_final_state, cu_seqlens`

### Registration
```python
# src/flag_gems/fused/FLA/__init__.py
from .chunk import chunk_gated_delta_rule_fwd

# src/flag_gems/fused/__init__.py
from flag_gems.fused.FLA import chunk_gated_delta_rule_fwd
```

### Hardware Adaptation
- NVIDIA RTX 4090: 68/68 tests passed, Triton path (K<=128), Native path (K>128)
- Iluvatar BI-V150: 68/68 tests passed, Native path (all K)

### Hybrid Dispatch Strategy

```python
# chunk.py - hybrid dispatch logic
K = k.shape[-1]
if _IS_ILUVATAR or K > 128:
    return _native_fwd(...)  # Native path
else:
    # Triton path for NVIDIA with K <= 128
    g, A = chunk_gated_delta_rule_fused_cumsum_kkt_solve_tril(...)
    w, u = recompute_w_u_fwd(...)
    h, v_new, final_state = chunk_gated_delta_rule_fwd_h(...)
    o = chunk_fwd_o(...)
```

### APIs Used

**Triton Path (NVIDIA, K<=128):**
- Triton JIT kernels with autotuning
- tl.exp, tl.dot, tl.make_block_ptr
- fp32 cast before exp() for fp16/bf16 support

**Native Path (Iluvatar or K>128):**
- torch.bmm, torch.matmul
- torch.linalg.solve_triangular
- torch.exp, torch.cumsum, torch.tril

### GPU Utilization Data (Iluvatar BI-V150)

**PyTorch Profiler Analysis** (shape: 1x512x8x64x64, dtype: float32):

| Kernel | CUDA Time (us) | % Total |
|--------|---------------|---------|
| elementwise (mul/sub/add) | 680 | 27.9% |
| solve_triangular (TRSM) | 401 | 16.5% |
| bmm (GEMM) | 359 | 14.7% |
| elementwise (sub) | 272 | 11.2% |
| copy_ | 239 | 9.8% |
| clamp | 159 | 6.5% |
| exp | 156 | 6.4% |
| matmul | 129 | 5.3% |
| add | 91 | 3.7% |
| **Total CUDA** | **2437** | **100%** |

### Test Coverage Matrix

| Test Category | Count | Coverage Dimensions |
|---------------|-------|---------------------|
| Basic functionality (batch) | 7 | B, T, H, K, V |
| cu_seqlens | 4 | Variable-length sequences |
| Multi-sequence packing | 2 | 2/3 sequences |
| GQA | 3 + 2 | Hg < H |
| Initial state | 1 + 1 | State management |
| Final state | 1 + 3 | State correctness |
| fp16/bf16 | 2 + 6 + 2 | Half precision |
| Non-contiguous tensor | 1 | Memory layout |
| Edge cases | 1 + 3 + 1 + 1 + 1 | Extreme values |
| Different K/V | 3 + 3 | Dimension combinations |
| Large batch | 4 | B=4,8 |
| Short sequences | 4 | T=1,8,16,32 |
| Many heads | 2 | H=16,32 |
| Precision characterization | 5 | T=64-1024 |
| Determinism | 1 | Reproducibility |
| **Total** | **68** | - |

### Test Function List (30 functions)

1. test_chunk_gated_delta_rule_batch (7)
2. test_chunk_gated_delta_rule_cu_seqlens (4)
3. test_chunk_gated_delta_rule_multi_sequence (1)
4. test_chunk_gated_delta_rule_gqa (3)
5. test_chunk_gated_delta_rule_with_initial_state (1)
6. test_chunk_gated_delta_rule_output_final_state (1)
7. test_chunk_gated_delta_rule_fp16_bf16 (2)
8. test_chunk_gated_delta_rule_non_contiguous (1)
9. test_chunk_gated_delta_rule_single_chunk (1)
10. test_chunk_gated_delta_rule_non_aligned (3)
11. test_chunk_gated_delta_rule_extreme_gates (1)
12. test_chunk_gated_delta_rule_zero_gates (1)
13. test_chunk_gated_delta_rule_kv_dims (3)
14. test_chunk_gated_delta_rule_large_batch (4)
15. test_chunk_gated_delta_rule_very_short (4)
16. test_chunk_gated_delta_rule_many_heads (2)
17. test_chunk_gated_delta_rule_gqa_ratios (2)
18. test_chunk_gated_delta_rule_multi_sequence_3 (1)
19. test_chunk_gated_delta_rule_cu_seqlens_non_aligned (1)
20. test_chunk_gated_delta_rule_cu_seqlens_gqa (1)
21. test_chunk_gated_delta_rule_final_state_correctness (3)
22. test_chunk_gated_delta_rule_beta_near_zero (1)
23. test_chunk_gated_delta_rule_beta_one (1)
24. test_chunk_gated_delta_rule_zero_v (1)
25. test_chunk_gated_delta_rule_deterministic (1)
26. test_chunk_gated_delta_rule_precision_characterization (5)
27. test_chunk_gated_delta_rule_fp16_bf16_shapes (6)
28. test_chunk_gated_delta_rule_cu_seqlens_half (2)
29. test_chunk_gated_delta_rule_initial_state_cu_seqlens (1)
30. test_chunk_gated_delta_rule_kv_extreme (3)

### Code Statistics
- `src/flag_gems/fused/FLA/chunk_gated_delta_rule_native.py`: 508 lines
- `src/flag_gems/fused/FLA/chunk.py`: 96 lines (modified: hybrid dispatch)
- `src/flag_gems/fused/FLA/wy_fast.py`: 1 line fix (fp16/bf16 exp)
- `src/flag_gems/fused/FLA/chunk_delta_h.py`: 12 lines fix (fp16/bf16 exp)
- `src/flag_gems/fused/FLA/chunk_o.py`: 2 lines fix (fp16/bf16 exp)
- `src/flag_gems/fused/FLA/fused_cumsum_kkt_solve_tril.py`: 1 line fix (fp16/bf16 exp)
- `tests/test_FLA/test_chunk_gated_delta_rule.py`: 1200+ lines (30 test functions, 68 cases)
- `benchmark/test_FLA/test_chunk_gated_delta_rule_perf.py`: 126 lines

### Documentation Completeness
- All functions have complete docstrings (description + parameter docs)
- Key algorithm steps have comments (state update, KKT computation, WY representation)
- Clear variable naming (h_state, g_exp, k_scaled_t, etc.)
- Complex logic has explanatory comments (e.g., why clamp(max=80.0))

### Code Structure
```python
def chunk_local_cumsum(g, chunk_size, cu_seqlens):     # Chunked cumulative sum
def chunk_scaled_dot_kkt_fwd(k, beta, g_cumsum, ...):  # KKT matrix computation
def solve_tril(A, cu_seqlens, output_dtype):            # Triangular solve (I-A)^{-1}
def recompute_w_u_fwd(k, v, beta, A, g_cumsum, ...):   # WY representation computation
def chunk_gated_delta_rule_fwd_fused(q, k, w, u, ...): # Fused output + state update
def chunk_gated_delta_rule_native_fwd(q, k, v, ...):   # Main entry function
```

## Modified Files

| File | Type | Description |
|------|------|-------------|
| `src/flag_gems/fused/FLA/chunk_gated_delta_rule_native.py` | New | PyTorch-native implementation (508 lines) |
| `src/flag_gems/fused/FLA/chunk.py` | Modified | Hybrid dispatch: Triton (K<=128) + Native (K>128) |
| `src/flag_gems/fused/FLA/wy_fast.py` | Modified | Fix fp16/bf16: cast to fp32 before exp() |
| `src/flag_gems/fused/FLA/chunk_delta_h.py` | Modified | Fix fp16/bf16: cast to fp32 before exp() |
| `src/flag_gems/fused/FLA/chunk_o.py` | Modified | Fix fp16/bf16: cast to fp32 before exp() |
| `src/flag_gems/fused/FLA/fused_cumsum_kkt_solve_tril.py` | Modified | Fix fp16/bf16: cast to fp32 before exp() |
| `tests/test_FLA/test_chunk_gated_delta_rule.py` | New | 30 unit tests (1200+ lines, 68 cases) |
| `benchmark/test_FLA/test_chunk_gated_delta_rule_perf.py` | New | FlagGems framework benchmark (126 lines) |

### NVIDIA RTX 4090
- GPU: NVIDIA GeForce RTX 4090 (24GB GDDR6X)
- PyTorch: 2.6.0+cu124
- CUDA: 12.4
- OS: Linux 6.8.0-107-generic

### Iluvatar BI-V150
- GPU: Iluvatar BI-V150 x 2 (32GB HBM)
- SDK: Iluvatar CoreX SDK 4.4.0.rc.11.20251201
- PyTorch: 2.7.1+corex.4.4.0
- Triton: 3.1.0+corex.4.4.0 (FlagTree 0.5.1+iluvatar3.1)

### CODE REVIEW

**Comment by douxetpur** [comment]
Hi, thank you for your participation in the FlagOS Open Computing Competition 🙏

After review, we have chosen to move forward with a different implementation for this operator. As a result, this PR will be closed.

We appreciate your time and contribution, and hope to see more of your submissions in the future.
